### PR TITLE
[improve][broker] Harden non-null checker

### LINKED ISSUE #17107: NPE when persistPositionToLedger
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Version

Master branch

### Minimal reproduce step

ReplicatorTest#testReplicatorWithFailedAck

### What did you expect to see?

No NEP.

### What did you see instead?

```
org.apache.bookkeeper.mledger.ManagedLedgerException: java.lang.NullPointerException
Caused by: java.lang.NullPointerException
	at java.util.Objects.requireNonNull(Objects.java:208) ~[?:?]
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl.persistPositionToLedger(ManagedCursorImpl.java:2814) ~[classes/:?]
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl.internalMarkDelete(ManagedCursorImpl.java:1968) ~[classes/:?]
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl.internalAsyncMarkDelete(ManagedCursorImpl.java:1911) ~[classes/:?]
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl.asyncDelete(ManagedCursorImpl.java:2226) ~[classes/:?]
	at org.apache.bookkeeper.mledger.impl.ManagedCursorImpl.asyncDelete(ManagedCursorImpl.java:2030) ~[classes/:?]
	at org.apache.pulsar.broker.service.persistent.PersistentReplicator.lambda$deleteFailed$4(PersistentReplicator.java:645) ~[classes/:?]
```

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by tisonkun**
This closes #17107.

Actually, after review all code path, I don't think there's a chance we pass `null` there now. But we can somehow harden the use case.

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 1]
```diff
@@ -3066,34 +3068,38 @@ boolean shouldCloseLedger(LedgerHandle lh) {
...
-    void switchToNewLedger(final LedgerHandle lh, final VoidCallback callback) {
+    void switchToNewLedger(@Nonnull LedgerHandle lh, VoidCallback callback) {
+        requireNonNull(lh, "LedgerHandle should not be null");
+
         if (log.isDebugEnabled()) {
             log.debug("[{}] Switching cursor {} to ledger {}", ledger.getName(), name, lh.getId());
         }
+
         persistPositionMetaStore(lh.getId(), lastMarkDeleteEntry.newPosition, lastMarkDeleteEntry.properties,
-                new MetaStoreCallback<Void>() {
```
Comment (BewareMyPower) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:3074]: Changes here are not necessary. Currently there is no strong code indent rule in the Java code. Please revert these changes.

**Review by BewareMyPower — CHANGES_REQUESTED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java (hunk 1 above)
Comment (tisonkun) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:3074]: Reverted.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java [hunk 2]
```diff
@@ -3066,21 +3068,25 @@ boolean shouldCloseLedger(LedgerHandle lh) {
         }
     }
 
-    void switchToNewLedger(final LedgerHandle lh, final VoidCallback callback) {
+    void switchToNewLedger(@Nonnull LedgerHandle lh, VoidCallback callback) {
```
Comment (BewareMyPower) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:3071]: I didn't get why should we add a `@Nonnull` annotation to `lh`. With this annotation, it seems that `lh` should not be null while `callback` might be `null`. However, `callback` should also be non-null.

Comment (BewareMyPower) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:3071]: I think we can close both.

**Comment by tisonkun** [comment]
Closing...

See https://github.com/apache/pulsar/pull/20096#discussion_r1169428743.
